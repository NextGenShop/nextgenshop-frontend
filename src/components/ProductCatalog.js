import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import Grid from "@material-ui/core/Grid";
import Card from "@material-ui/core/Card";
import CardMedia from "@material-ui/core/CardMedia";
import CardContent from "@material-ui/core/CardContent";
import Button from "@material-ui/core/Button";
import mockProducts from "../store/mockdata/MockSupermarketDataset.json";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableContainer from "@material-ui/core/TableContainer";
import TableRow from "@material-ui/core/TableRow";
import Tooltip from "@material-ui/core/Tooltip";

const useStyles = makeStyles((theme) => ({
  root: {
    height: "29vh",
    overflowY: "scroll",
  },
  tooltip: {
    backgroundColor: "#f5f5f9",
    color: "rgba(0, 0, 0, 0.87)",
    maxWidth: 300,
    border: "1px solid #dadde9",
  },
  card: {
    display: "flex",
  },
  details: {
    display: "flex",
    flexDirection: "column",
  },
  content: {
    flex: "1 0 auto",
  },
  buttons: {
    display: "flex",
    alignItems: "center",
    paddingLeft: theme.spacing(1),
    paddingBottom: theme.spacing(1),
  },
  media: {
    width: 80,
  },
  productText: {
    fontWeight: "bold",
    fontSize: "13px",
  },
  smallText: {
    fontSize: "12px",
  },
  mb: {
    marginBottom: theme.spacing(2),
  },
}));

export default function ProductCatalog({ addToBasket, searchQuery, retailer, limit, tableView }) {
  const classes = useStyles();

  const getProducts = () => {
    let products = mockProducts.filter((product) => product.retailer === retailer);
    if (searchQuery)
      products = products.filter(
        (product) =>
          product.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
          product.productId.toString() === searchQuery
      );
    if (limit) products = products.slice(0, Math.min(limit, products.length));
    return products;
  };

  return tableView ? (
    <TableContainer className={classes.root}>
      <Table size="small" aria-label="product catalog table">
        <TableBody>
          {getProducts().map((product) => (
            <TableRow key={product.productId}>
              <Tooltip title={<img alt="product" className={classes.media} src={product.image} />}>
                <TableCell component="th" scope="row">
                  {product.name}
                </TableCell>
              </Tooltip>
              <TableCell align="right">£ {product.price.toFixed(2)}</TableCell>
              <TableCell align="right">
                <Button
                  variant="contained"
                  size="small"
                  color="primary"
                  onClick={() => addToBasket(product)}
                >
                  Add to basket
                </Button>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  ) : (
    <Grid container spacing={2} className={classes.root}>
      {getProducts().map((product) => (
        <Grid item md={4}>
          <Card className={classes.card}>
            <div className={classes.details}>
              <CardContent className={classes.content}>
                <Typography className={classes.productText} gutterBottom variant="subtitle2">
                  {product.name}
                </Typography>
                <Typography
                  className={classes.smallText}
                  display="inline"
                  variant="body2"
                  color="textSecondary"
                  component="p"
                >
                  Price:{" "}
                </Typography>
                <Typography
                  className={classes.smallText}
                  display="inline"
                  variant="body2"
                  color="primary"
                  component="p"
                >
                  £ {product.price.toFixed(2)}
                </Typography>
                <br />
                <Typography
                  className={classes.smallText}
                  display="inline"
                  variant="body2"
                  color="textSecondary"
                  component="p"
                >
                  Retailer:{" "}
                </Typography>
                <Typography
                  className={classes.smallText}
                  display="inline"
                  variant="body2"
                  color="primary"
                  component="p"
                >
                  {product.retailer}
                </Typography>
              </CardContent>
              <div className={classes.buttons}>
                <Button
                  variant="contained"
                  size="small"
                  color="primary"
                  onClick={() => addToBasket(product)}
                >
                  Add to basket
                </Button>
              </div>
            </div>
            <CardMedia className={classes.media} image={product.image} />
          </Card>
        </Grid>
      ))}
    </Grid>
  );
}
