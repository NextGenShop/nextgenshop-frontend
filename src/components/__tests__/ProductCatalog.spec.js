import React from 'react';
import { shallow } from 'enzyme';
import { ProductCatalog, mapDispatchToProps } from '../ProductCatalog';
import mockBasketItems from '../../store/mock/MockBasketItems.json';
import mockProducts from '../../store/mock/MockSupermarketDataset.json';

jest.mock('../../store/api/basket', () => {
  return {
    ...jest.requireActual('../../store/api/basket'),
    actions: {
      updateBasket: () => 'UPDATE_BASKET',
    },
  };
});

jest.mock('../../store/api/product', () => {
  return {
    ...jest.requireActual('../../store/api/product'),
    actions: {
      getProducts: () => 'GET_PRODUCTS',
    },
  };
});

describe('ProductCatalog', () => {
  const defaultProps = {
    searchQuery: 'mockQuery',
    retailer: 'mockRetailer',
    limit: 9,
    tableView: false,
    products: [],
    shopperId: 0,
    dispatchGetProducts: jest.fn(),
    dispatchUpdateBasket: jest.fn(),
    basket: { items: mockBasketItems, totalPrice: 0, shopper: 0 },
  };

  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });

  test('should correctly render with props', () => {
    const component = shallow(<ProductCatalog {...defaultProps} />);
    expect(component).toMatchSnapshot();
  });

  test('should correctly render with products', () => {
    const component = shallow(
      <ProductCatalog
        {...defaultProps}
        products={[mockProducts[0], mockProducts[1]]}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test('should correctly fetch products on render', () => {
    shallow(<ProductCatalog {...defaultProps} />);
    expect(defaultProps.dispatchGetProducts).toBeCalledTimes(1);
  });

  test('should correctly render table view', () => {
    const component = shallow(
      <ProductCatalog {...defaultProps} tableView={true} />
    );
    expect(component).toMatchSnapshot();
  });

  test('should correctly render table view with products', () => {
    const component = shallow(
      <ProductCatalog
        {...defaultProps}
        tableView={true}
        products={[mockProducts[0], mockProducts[1]]}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test('should correctly add a product to basket', () => {
    const component = shallow(
      <ProductCatalog {...defaultProps} products={[mockProducts[0]]} />
    );
    component
      .find('[aria-label="addToBasketButton"]')
      .first()
      .simulate('click');
    expect(defaultProps.dispatchUpdateBasket).toBeCalledTimes(1);
  });

  test('should correctly add a product to basket in table view', () => {
    const component = shallow(
      <ProductCatalog
        {...defaultProps}
        tableView={true}
        products={[mockProducts[0]]}
      />
    );
    component
      .find('[aria-label="addToBasketButton"]')
      .first()
      .simulate('click');
    expect(defaultProps.dispatchUpdateBasket).toBeCalledTimes(1);
  });

  test('should dispatch correct actions', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).dispatchUpdateBasket();
    mapDispatchToProps(dispatch).dispatchGetProducts();
    expect(dispatch.mock.calls[0][0]).toBe('UPDATE_BASKET');
    expect(dispatch.mock.calls[1][0]).toBe('GET_PRODUCTS');
  });
});
