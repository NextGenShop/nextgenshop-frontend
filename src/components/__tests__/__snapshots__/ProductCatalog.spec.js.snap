// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCatalog should correctly render table view 1`] = `
<WithStyles(ForwardRef(TableContainer))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Table))
    aria-label="product catalog table"
    size="small"
  >
    <WithStyles(ForwardRef(TableBody)) />
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>
`;

exports[`ProductCatalog should correctly render table view with products 1`] = `
<WithStyles(ForwardRef(TableContainer))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Table))
    aria-label="product catalog table"
    size="small"
  >
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(ForwardRef(TableRow))
        key="0"
      >
        <WithStyles(ForwardRef(Tooltip))
          title={
            <img
              alt="product"
              className="makeStyles-media-7"
              src="placeholder_image.png"
            />
          }
        >
          <WithStyles(ForwardRef(TableCell))
            component="th"
            scope="row"
          >
            Milk Chocolate Biscuits 9 Pack
          </WithStyles(ForwardRef(TableCell))>
        </WithStyles(ForwardRef(Tooltip))>
        <WithStyles(ForwardRef(TableCell))
          align="right"
        >
          £ 
          1.50
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          align="right"
        >
          <WithStyles(ForwardRef(Button))
            aria-label="addToBasketButton"
            color="primary"
            onClick={[Function]}
            size="small"
            variant="contained"
          >
            Add to basket
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        key="1"
      >
        <WithStyles(ForwardRef(Tooltip))
          title={
            <img
              alt="product"
              className="makeStyles-media-7"
              src="placeholder_image.png"
            />
          }
        >
          <WithStyles(ForwardRef(TableCell))
            component="th"
            scope="row"
          >
            Cookie Dough Vanilla Ice Cream 465Ml
          </WithStyles(ForwardRef(TableCell))>
        </WithStyles(ForwardRef(Tooltip))>
        <WithStyles(ForwardRef(TableCell))
          align="right"
        >
          £ 
          4.50
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          align="right"
        >
          <WithStyles(ForwardRef(Button))
            aria-label="addToBasketButton"
            color="primary"
            onClick={[Function]}
            size="small"
            variant="contained"
          >
            Add to basket
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>
`;

exports[`ProductCatalog should correctly render with products 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    key="0"
    md={4}
  >
    <WithStyles(ForwardRef(Card))
      className="makeStyles-card-3"
    >
      <div
        className="makeStyles-details-4"
      >
        <WithStyles(ForwardRef(CardContent))
          className="makeStyles-content-5"
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-productText-8"
            gutterBottom={true}
            variant="subtitle2"
          >
            Milk Chocolate Biscuits 9 Pack
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="textSecondary"
            component="p"
            display="inline"
            variant="body2"
          >
            Price:
             
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="primary"
            component="p"
            display="inline"
            variant="body2"
          >
            £ 
            1.50
          </WithStyles(ForwardRef(Typography))>
          <br />
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="textSecondary"
            component="p"
            display="inline"
            variant="body2"
          >
            Retailer:
             
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="primary"
            component="p"
            display="inline"
            variant="body2"
          >
            Mock Retailer A
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(CardContent))>
        <div
          className="makeStyles-buttons-6"
        >
          <WithStyles(ForwardRef(Button))
            aria-label="addToBasketButton"
            className="makeStyles-addButton-11"
            color="primary"
            onClick={[Function]}
            size="small"
            variant="contained"
          >
            Add
          </WithStyles(ForwardRef(Button))>
        </div>
      </div>
      <WithStyles(ForwardRef(CardMedia))
        className="makeStyles-media-7"
        image="placeholder_image.png"
      />
    </WithStyles(ForwardRef(Card))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    key="1"
    md={4}
  >
    <WithStyles(ForwardRef(Card))
      className="makeStyles-card-3"
    >
      <div
        className="makeStyles-details-4"
      >
        <WithStyles(ForwardRef(CardContent))
          className="makeStyles-content-5"
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-productText-8"
            gutterBottom={true}
            variant="subtitle2"
          >
            Cookie Dough Vanilla Ice Cream 465Ml
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="textSecondary"
            component="p"
            display="inline"
            variant="body2"
          >
            Price:
             
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="primary"
            component="p"
            display="inline"
            variant="body2"
          >
            £ 
            4.50
          </WithStyles(ForwardRef(Typography))>
          <br />
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="textSecondary"
            component="p"
            display="inline"
            variant="body2"
          >
            Retailer:
             
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-smallText-9"
            color="primary"
            component="p"
            display="inline"
            variant="body2"
          >
            Mock Retailer A
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(CardContent))>
        <div
          className="makeStyles-buttons-6"
        >
          <WithStyles(ForwardRef(Button))
            aria-label="addToBasketButton"
            className="makeStyles-addButton-11"
            color="primary"
            onClick={[Function]}
            size="small"
            variant="contained"
          >
            Add
          </WithStyles(ForwardRef(Button))>
        </div>
      </div>
      <WithStyles(ForwardRef(CardMedia))
        className="makeStyles-media-7"
        image="placeholder_image.png"
      />
    </WithStyles(ForwardRef(Card))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`ProductCatalog should correctly render with props 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
  spacing={2}
/>
`;
